crates/holochain_p2p/src/hc_report.rs

```diff
@@ -0,0 +1,599 @@
+                path: "/tmp/hc-report".into(),
+                fetched_op_interval_s: 60.0,
+            }
+        }
+    }
+
+    /// Module-level configuration for HcReport.
+    #[derive(Debug, Default, Clone, serde::Serialize, serde::Deserialize)]
+    #[serde(rename_all = "camelCase")]
+    pub struct HcReportModConfig {
+        /// HcReport configuration.
+        pub hc_report: HcReportConfig,
+    }
+}
+
+pub use config::*;
+
+/// A default no-op report module.
+#[derive(Debug)]
+pub struct HcReportFactory {}
+
+impl HcReportFactory {
+    /// Construct a new [`HcReportFactory`]
+    pub fn create() -> DynReportFactory {
+        let out: DynReportFactory = Arc::new(Self {});
+        out
+    }
+}
+
+impl ReportFactory for HcReportFactory {
+    fn default_config(&self, config: &mut Config) -> K2Result<()> {
+        config.set_module_config(&HcReportModConfig::default())?;
+        Ok(())
+    }
+
+    fn validate_config(&self, _config: &Config) -> K2Result<()> {
+        Ok(())
+    }
+
+    fn create(
+        &self,
+        builder: Arc<Builder>,
+        tx: DynTransport,
+    ) -> BoxFut<'static, K2Result<DynReport>> {
+        Box::pin(async move {
+            let config: HcReportModConfig = builder.config.get_module_config()?;
+            let out: DynReport = HcReport::create(config.hc_report, tx)?;
+            Ok(out)
+        })
+    }
+}
+
+/// Type sent on our internal command channel.
+enum Cmd {
+    /// Indicates we have received op data from a remote peer.
+    FetchedOp {
+        space_id: SpaceId,
+        source: Url,
+        size_bytes: u64,
+    },
+}
+
+struct HcReport {
+    /// Weak self reference.
+    this: Weak<Self>,
+
+    /// Timing loop task.
+    task: tokio::task::AbortHandle,
+
+    /// Map of LocalAgentStores letting us sign reports.
+    spaces: Mutex<HashMap<SpaceId, DynLocalAgentStore>>,
+
+    /// Sender side of command channel.
+    cmd_send: tokio::sync::mpsc::Sender<Cmd>,
+
+    /// Receiver side of command channel.
+    cmd_recv: Mutex<tokio::sync::mpsc::Receiver<Cmd>>,
+
+    /// Transport for sending messages to remote peers.
+    tx: DynTransport,
+
+    /// The log file writer.
+    file_writer: Mutex<tracing_appender::non_blocking::NonBlocking>,
+
+    /// The guard that shuts down the non-blocking task on drop.
+    _file_guard: tracing_appender::non_blocking::WorkerGuard,
+}
+
+impl Drop for HcReport {
+    fn drop(&mut self) {
+        self.task.abort();
+    }
+}
+
+impl std::fmt::Debug for HcReport {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("HcReport").finish()
+    }
+}
+
+impl TxBaseHandler for HcReport {}
+
+impl TxModuleHandler for HcReport {
+    fn recv_module_msg(
+        &self,
+        _peer: Url,
+        _space_id: SpaceId,
+        _module: String,
+        data: bytes::Bytes,
+    ) -> K2Result<()> {
+        // on receiving a report, write it to our report file
+        self.write_bytes(&data);
```

Comment:
Are there any limits on this?

This is the downside of it being a separate module. It's not tied to the requests and responses of the fetch module.